Source/Core/VideoCommon/TextureConverterShaderGen.cpp

```diff
@@ -0,0 +1,247 @@
+  uid_data->scale_by_half = scale_by_half;
+
+  return out;
+}
+
+ShaderCode GenerateTextureConverterShaderCode(APIType api_type,
+                                              const convertion_shader_uid_data* uid_data)
+{
+  ShaderCode out;
+
+  out.Write(
+      "SAMPLER_BINDING(9) uniform sampler2DArray samp9;\n"
+      "in vec3 f_uv0;\n"
+      "out vec4 ocol0;\n"
+      "void main(){\n"
+      "	vec4 texcol = texture(samp9, %s);\n",
+      uid_data->is_depth_copy ?
+          (g_ActiveConfig.bStereoEFBMonoDepth ? "vec3(f_uv0.xy, 0.0)" : "vec3(f_uv0.xy, f_uv0.z)") :
+          "f_uv0");
+
+  if (uid_data->is_depth_copy)
+  {
+    out.Write("	int depth = int(texcol.x * 16777216.0);\n"
+
+              // Convert to Z24 format
+              "	ivec4 workspace;\n"
+              "	workspace.r = (depth >> 16) & 255;\n"
+              "	workspace.g = (depth >> 8) & 255;\n"
+              "	workspace.b = depth & 255;\n"
+
+              // Convert to Z4 format
+              "	workspace.a = (depth >> 16) & 0xF0;\n"
+
+              // Normalize components to [0.0..1.0]
+              "	texcol = vec4(workspace) / 255.0;\n");
+    switch (uid_data->dst_format)
+    {
+    case EFBCopyFormat::R4:  // Z4
+      out.Write("	ocol0 = texcol.aaaa;\n");
+      break;
+
+    case EFBCopyFormat::R8_0x1:  // Z8
+    case EFBCopyFormat::R8:      // Z8H
+      out.Write("	ocol0 = texcol.rrrr;\n");
+      break;
+
+    case EFBCopyFormat::RA8:  // Z16
+      out.Write("	ocol0 = texcol.gggr;\n");
+      break;
+
+    case EFBCopyFormat::RG8:  // Z16 (reverse order)
+      out.Write("	ocol0 = texcol.rrrg;\n");
+      break;
+
+    case EFBCopyFormat::RGBA8:  // Z24X8
+      out.Write("	ocol0 = vec4(texcol.rgb, 1.0);\n");
+      break;
+
+    case EFBCopyFormat::G8:  // Z8M
+      out.Write("	ocol0 = texcol.gggg;\n");
+      break;
+
+    case EFBCopyFormat::B8:  // Z8L
+      out.Write("	ocol0 = texcol.bbbb;\n");
+      break;
+
+    case EFBCopyFormat::GB8:  // Z16L - copy lower 16 depth bits
+      // expected to be used as an IA8 texture (upper 8 bits stored as intensity, lower 8 bits
+      // stored as alpha)
+      // Used e.g. in Zelda: Skyward Sword
+      out.Write("	ocol0 = texcol.gggb;\n");
+      break;
+
+    default:
+      ERROR_LOG(VIDEO, "Unknown copy zbuf format: 0x%X", static_cast<int>(uid_data->dst_format));
+      out.Write("	ocol0 = vec4(texcol.bgr, 0.0);\n");
+      break;
+    }
+  }
+  else if (uid_data->is_intensity)
+  {
+    std::array<float, 28> colmat = {};
+    float* const const_add = &colmat[16];
+    float* const color_mask = &colmat[20];
+    color_mask[0] = color_mask[1] = color_mask[2] = color_mask[3] = 255.0f;
+    color_mask[4] = color_mask[5] = color_mask[6] = color_mask[7] = 1.0f / 255.0f;
+
+    const_add[0] = const_add[1] = const_add[2] = 16.0f / 255.0f;
+
+    switch (uid_data->dst_format)
+    {
+    case EFBCopyFormat::R4:      // I4
+    case EFBCopyFormat::R8_0x1:  // I8
+    case EFBCopyFormat::R8:      // IA4
```

Comment:
Sorry, I should have been specific. R8 is not IA4. Other cases have mismatched comments.